### DISCUSSION TITLE
フルスタックNext.jsアプリのデプロイ先を決める

### DISCUSSION THREAD

**Discussion by kazzyfrog** [lgtm-factory/lgtm-factory#60] [Scrap]
## 概要

表題の通り、デプロイ先を決めます！

それに関するリサーチを、つらつら書いて、選定していくスレです！

---

要件として、
画像のURLを返すAPIを建てる必要があるので、サーバーサイドでの処理が必要です。
（静的サイトではない）

なので、GitHub Pagesは、対象から外れる、、！
（確か、静的サイトのみだったはず。。！後で再度調べる）

**Comment by kazzyfrog** [comment]
vercelか、Cloudflare Pages or Workers がいいのかな結局。。
比較する！

**Reply by kazzyfrog** [reply]
Vercel
[公式/制限](https://vercel.com/docs/limits/overview)

**Comment by kazzyfrog** [comment]
フルスタックNext.jsアプリの場合、サーバーサイドの処理は、Workersで動くので、同じっぽい！


> Cloudflare Workersの他にPagesというのがありまして、一般的にはNext.jsやAstroなどのフレームワークを用いて「フルスタック」なアプリケーションを作るプラットフォームです。 以前はWorkersとPagesは明確に分かれていたのですが、それが統合されつつあります。
Pagesはもともと静的なページをホストするのが主なユースケースだったのが、「SSR」というキーワードから分かる通り、ダイナミックな処理が増えてきました。 そのPagesのダイナミック処理はWorkersで動いているので、その境界線が曖昧になるんですね。
実際、ダッシュボードではWorkersもPagesも同じセクションになりました。

[参考](https://yusukebe.com/posts/2024/cloudflare-workers-updates/)

**Reply by kazzyfrog** [reply]
無料枠の制限は、載ってない、、っぽいけど、多分Workersと同じ？

そしたら、
> 1日あたり10万件のリクエストを含む

[公式](https://www.cloudflare.com/ja-jp/plans/developer-platform/)

**Reply by kazzyfrog** [reply]
Cloudflare Pagesの利点は、ざっくり：
- 複数人での管理も無料
- 商用利用可能？

**Comment by kazzyfrog** [comment]
決定的な差がなければ、一旦vercelでもいいもかも！
簡単だし！あとで移行できるし！

**Reply by kazzyfrog** [reply]
あ、でも移行してURL変わったら、画像のパス全部変わるのか。。
そしたら、既存の画像が表示されなくなる。

なので、できれば移行したくない、、感じか！

**Reply by kazzyfrog** [reply]
回避すべきはドメイン、URLが変わること、、なのか？

- 現時点でしっかり比較検討して、決めるか、
- とりあえずvercelにデプロイして、制限？などの通知が来たら、移行する。

移行した場合、ドメイン変わる。。

**Reply by kazzyfrog** [reply]
Cloudflare Pages なのか、、、？

**Comment by kazzyfrog** [comment]
個人開発だとしたら、
- ドメイン新規で取得するか、
- すでに持ってるドメインのサブドメイン割り当てるか

なので、デプロイ先で、もし制限などの問題あっても、後から移行できるので、一旦vercelで良さそう！

**Reply by kazzyfrog** [reply]
無料のドメイン取る手もあるけど、あまりイケてない。

APIとフロントエンドアプリを、切り離すのは大変、、というかそこまでする必要ない。。！

**Comment by kazzyfrog** [comment]
## 一旦整理

要求：
- Next.jsのフルスタックアプリをデプロイする
    - URLの値に応じて、画像が生成され、返却される

そして、APIリクエスト？の制限と言われても、感覚的にどれくらいの余裕があるのか、なのかわかっていない。。
画像毎なので、自分たちの普段使いでも、たくさんAPIリクエストが飛ぶことにはなる、、？とは、思う！

なので、
1. 無料で、１番利用枠が広い（APIサーバーへのリクエスト）ところを探す
2. 万が一、制限を超えた・超えそうな時の対処を考えておく
3. その場合、できれば、ドメインは変更したくない

**Reply by kazzyfrog** [reply]
デプロイ先を考えるにあたって、
サイトのトラフィックが少ない場合は、考えなくて良い。

なぜなら、どこにデプロイでも問題が発生しないから。

なので、考えるべきは、トラフィック（リクエスト）多くなった場合に、どうするか！

**Comment by kazzyfrog** [comment]
Claude に聞いた

Vercelと Cloudflare Pages の詳細な比較を行い、選定のための情報を提供します。

1. パフォーマンスと CDN:
   - Vercel: グローバルエッジネットワークを使用。高速なコンテンツ配信が可能。
   - Cloudflare Pages: Cloudflare の広大な CDN ネットワークを利用。世界中で高速なコンテンツ配信が可能。

2. サーバーレス関数 (API) の制限:
   - Vercel: 無料プランで 100GB-時間/月
   - Cloudflare Pages: Workers を使用し、無料プランで 100,000 リクエスト/日

3. 帯域幅:
   - Vercel: 無料プランで 100GB/月
   - Cloudflare Pages: 制限なし

4. ビルド時間:
   - Vercel: 無料プランで 6000 分/月
   - Cloudflare Pages: 無料プランで 500 ビルド/月

5. Next.js との互換性:
   - Vercel: Next.js の開発元であり、最適化されている
   - Cloudflare Pages: Next.js をサポートしているが、一部機能に制限がある可能性がある

6. カスタムドメイン:
   - Vercel: 無料プランでも利用可能
   - Cloudflare Pages: 無料プランでも利用可能

7. 環境変数:
   - Vercel: 豊富な環境変数管理機能
   - Cloudflare Pages: 環境変数のサポートあり

8. プレビュー機能:
   - Vercel: プルリクエストごとのプレビューデプロイメント
   - Cloudflare Pages: プレビューデプロイメントをサポート

9. 有料プラン:
   - Vercel: $20/月から
   - Cloudflare Pages: $5/月から（一部機能は別料金）

選定:

1. トラフィックが多い場合や、帯域幅の制限を気にする場合:
   Cloudflare Pages が優位です。無制限の帯域幅は大きな利点となります。

2. Next.js の最適化と完全な互換性を重視する場合:
   Vercel が優位です。Next.js の開発元であり、最適な環境を提供します。

3. コスト面を重視する場合:
   Cloudflare Pages の方が安価な有料プランを提供しています。

4. API リクエストの頻度が高い場合:
   日単位で考えると Cloudflare Pages の 100,000 リクエスト/日が有利ですが、月単位では Vercel の 100GB-時間/月も十分な場合があります。使用パターンによって判断が分かれます。

5. 開発者体験とエコシステムを重視する場合:
   Vercel は Next.js に特化した機能や統合が充実しているため、優位です。

結論:
Next.js アプリケーションの開発では、Vercel の方が総合的に優れています。特に、Next.js との互換性、最適化、開発者体験の面で大きな利点があります。

ただし、プロジェクトが大量のトラフィックを扱う可能性がある場合や、コスト面での制約が厳しい場合は、Cloudflare Pages も十分に検討に値します。

最終的な選択は、プロジェクトの具体的な要件、予算、そして将来のスケールアップの可能性を考慮して行うことをお勧めします。両プラットフォームともに無料プランがあるので、実際に小規模なプロジェクトをデプロイして比較してみるのも良いでしょう。

**Reply by kazzyfrog** [reply]
追加：

---

next/ImageResponseを使用した動的な画像生成についての具体的な使用例をお聞きして、より適切な判断ができます。この場合の考察は以下のとおりです：

1. 処理の性質:
   next/ImageResponseを使用した画像生成は、一般的に以下の特徴があります：
   - 比較的軽量な処理：単純なテキストや基本的な図形の描画の場合
   - 中程度の処理：複雑なレイアウトや多数の要素を含む場合
   - 重い処理：高解像度や複雑なグラフィックス処理を含む場合

2. 実行時間とリソース使用:
   - 多くの場合、数百ミリ秒から数秒程度で処理が完了すると予想されます。
   - メモリ使用量は通常それほど大きくありませんが、画像のサイズや複雑さによって変わります。

3. Vercel vs Cloudflare Pages:

   Vercel:
   - Next.jsの開発元であり、next/ImageResponseに最適化されている可能性が高い
   - 処理時間とメモリ使用量に基づく柔軟な計算方式（100GB-時間/月）
   - Edge Functionsを使用すれば、より高速な処理が可能

   Cloudflare Pages:
   - 単純なリクエスト数ベースの制限（100,000リクエスト/日）
   - Workers KVやDurable Objectsなど、追加のストレージオプションが利用可能

4. 判断:
   この場合、Vercelの方が適していると考えられます。理由は以下の通りです：
   - Next.jsとの互換性が高く、最適化されている
   - 画像生成処理は比較的軽量で短時間で完了する可能性が高いため、Vercelの柔軟な計算方式で効率的に処理できる
   - Edge Functionsを使用することで、より高速なレスポンスが可能

ただし、以下の点に注意が必要です：
- 生成する画像の複雑さや頻度によっては、リソース使用量が増加する可能性がある
- 高トラフィックの場合、Cloudflareの無制限帯域幅が魅力的になる可能性がある

結論：
next/ImageResponseを使用した画像生成の場合、多くのシナリオでVercelが適していると考えられます。ただし、具体的な使用パターンや予想されるトラフィック量によっては、実際にテストデプロイを行い、両プラットフォームでのパフォーマンスと制限を比較することをお勧めします。

最終的な選択は、アプリケーションの具体的な要件、期待されるトラフィック量、そして将来のスケーリング計画に基づいて行うのが賢明です。

**Comment by kazzyfrog** [comment]
無料枠のリクエスト自体は、vercelの方が余裕あるっぽい。。！？
next/ImageResponseを使用した画像生成の場合、ほとんど軽量な処理だろうし！

---

ご質問ありがとうございます。ざっくりとした比較をするために、いくつかの仮定を置いて計算してみましょう。

1. Cloudflare Pages: 100,000リクエスト/日
   月に換算すると: 100,000 * 30 = 3,000,000リクエスト/月

2. Vercel: 100GB-時間/月
   これを理解しやすいリクエスト数に変換するには、平均的なリクエストの実行時間とメモリ使用量を仮定する必要があります。

仮定：
- 平均的なリクエストが0.1秒（100ミリ秒）で処理される
- 平均的なメモリ使用量が128MB（0.125GB）

計算：
100GB-時間 = 100 * 3600秒 = 360,000GB-秒
1リクエストあたりのGB-秒 = 0.1秒 * 0.125GB = 0.0125GB-秒
可能なリクエスト数 = 360,000GB-秒 / 0.0125GB-秒 = 28,800,000リクエスト/月

この仮定の下では：
Vercel: 約28,800,000リクエスト/月
Cloudflare Pages: 約3,000,000リクエスト/月

したがって、この仮定の下では、Vercelの枠の方が大きいと言えます。

ただし、以下の点に注意が必要です：
1. 実際のリクエストの処理時間やメモリ使用量は、アプリケーションの性質によって大きく異なります。
2. Cloudflareの制限は日単位なので、短期的なスパイクに強い可能性があります。
3. Vercelの制限は月単位で、より長期的な使用パターンに適しています。

結論：
一般的なケースでは、Vercelの枠の方が大きい可能性が高いですが、実際の使用状況によって変わる可能性があります。特に、リクエストが非常に軽量で短時間で処理される場合、Vercelの利点はさらに大きくなります。一方、リクエストが複雑で長時間の処理を要する場合、Cloudflareの方が有利になる可能性があります。

実際のアプリケーションでの使用量を測定し、それに基づいて判断することをお勧めします。

**Comment by kazzyfrog** [comment]
ここまでまとめ：
- 無料プランなら、vercelが良さそう
- もし有料プランになるなら、Cloudflare Pages

多分無料使い切ることないだろうけど、、Cloudflare Pagesは有料でも5ドル。
その時は、商用できるので、広告貼る、、？GitHubスポンサー貼る？とかで、ペイしそうな気もする！

**Reply by kazzyfrog** [reply]
そして、万が一、移行することを考えたら、独自ドメインを使えたら完璧！
心配事がなくなる！

**Reply by kagomen** [reply]
とりあえずCloudflareにデプロイして、あまりにも開発者体験が悪かったらVercelにデプロイする、みたいなのはどうなんでしょうか？👀
デプロイ先変えるだけで、ソースコードの中身も結構変えなきゃいけない感じなんですかね・・・？

**Reply by kazzyfrog** [reply]
確かに、それでいいかもですね！

1日あたり10万件のリクエストまで、、耐えれるので！w

> ソースコードの中身も結構変えなきゃいけない感じ

これは、少しで大丈夫だと思います！！
Cloudflareでも結構対応したいたので！

**Reply by kagomen** [reply]
良いですね！✨

個人的に、CloudflareがどのくらいNext.jsに対応してるのかを知れる機会になるので、嬉しいかも！

**Reply by kazzyfrog** [reply]
確かに、vercelから移行するかもしれないコスト考えたら、初めからCloudflareがいいかもですね！👍

**Comment by kazzyfrog** [comment]
## 結論

Cloudflare Pages にデプロイする

サービス名を変更する可能性が高そう ＆ 最悪本リリースまでにデプロイ先変えることもできるように、
開発用の環境で、仮URL（`lgtm-factory.pages.dev`）でいいかも！

本リリース？時に、確定したサービス名のドメインで変更する形で！

一旦、Next.jsのセットアップと、開発用でデプロイまでする！🚀

**Reply by kazzyfrog** [reply]
noteで書いていたような、このコミュニティ（アクセラ）のことを掲載していく、
`b13o.com`のドメインを持っているので、サブドメインでそれを使っても良いし！

多分、移行すること考える心配はないと思うけど、、！
